### DISCUSSION TITLE
Absorbing cia402.comp into lcec cia402 drivers.

### DISCUSSION THREAD

**Discussion by rodw-au** [linuxcnc-ethercat/linuxcnc-ethercat#424] [Ideas]
There is  alot I still don't understand about the lcec_cia402 drivers and its derivitives. I am a long way short of writing driver code. I am going to assume that lcec_cia402_write() and lcec_cia402_read() can see and act on all drive parameters without resorting to pins. The read and write code in the cia402.comp has a natural home in the lcec driver

Please correct me if I am wrong.

If this is the case, than there is no need to replicate all of the pins created by cia402.comp if it is absorbed into the lcec driver and directly accessed. Technically, the only additional pins are required are to map to the three homing pins in cia402_homecomp.comp
ref: https://github.com/rodw-au/cia402_homecomp

```
//Homing IOs
pin io bit home         "true starts the Drives internal home procedure, connect this to joint.x.index-enable pin";
pin out bit stat_homed  "true indicates that the Drive is internaly homed";
pin out bit stat_homing "true indicates that the Drives homing procedure is running";
```
I also dislike the extensive use of bitmapped maths to parse the controlword and statusword values. Our job as programmers should be to eliminate obfuscation and simplify the user's understanding.

I would like to suggest the following syntax be used to deal with these bitmapped fields 

```
//CIA Control Word
typedef union
{
  struct
  {
    unsigned char SwitchOn        : 1;  // 00
    unsigned char EnableVoltage   : 1;  // 01
    unsigned char QuickStop       : 1;  // 02
    unsigned char EnableOperation : 1;  // 03
    unsigned char StartHome       : 1;  // 04 not used in CSP or homing
    unsigned char bit5            : 1;  // 05 not used in CSP or homing
    unsigned char bit6            : 1;  // 06 not used in CSP or homing
    unsigned char Faultreset      : 1;  // 07
    unsigned char Timeout         : 1;  // 08
    unsigned char bit9            : 1;  // 09 not used in any mode
    unsigned char keep            : 6;  // 10-15 not used in any mode
  }b;
    hal_u32_t Word;
}Control_t;

//CIA Status Word
typedef union
{
  struct
  {
    unsigned char ReadyToSwitchOn  : 1;  // 00
    unsigned char SwitchOn         : 1;  // 01
    unsigned char OperationEnabled : 1;  // 02
    unsigned char Fault            : 1;  // 03
    unsigned char VoltageEnabled   : 1;  // 04
    unsigned char QuickStop        : 1;  // 05
    unsigned char SwitchOnDisabled : 1;  // 06
    unsigned char Warning          : 1;  // 07
    unsigned char keep1            : 1;  // 08
    unsigned char Remote           : 1;  // 09
    hal_bit_t     TargetReached    : 1;  // 10
    unsigned char bit11            : 1;  // 11
    unsigned char bit12            : 1;  // 12
    unsigned char keep2            : 3;  // 13-15
  }b;
    hal_u32_t Word;
}Status_t;

    Status_t     Status;  		// IN  - CIA Status Word with bitmap (not a pin)
    Control_t    Control; 		// OUT - CIA Control Word with bitmap (not a pin)
```

This then gives a much clearer syntax with zero overhead because of the use of C's union typedef

```
stat_target_reached = Status.b.TargetReached;
stat_word=Status.Word;
```
Hopefully the cia402.comp can be migrated quite easilly. We already know that it is unchanged across drives so perhaps it can be called from  seperate procedure from all of the cia402.x devices.

I am happy to help and discuss ideas as i did make a good go  at moving this component previously
https://github.com/rodw-au/ciahome/tree/main

**Comment by scottlaird** [comment]
Let's see if I can respond in order here:

- The `_read` and `_write` functions can do pretty much anything, although from a timing perspective they probably shouldn't depend on doing SDO reads.  Anything that's already in memory (PDOs and internal variables of whatever sort) or can be queued (SDO writes) is fine.
- I agree that setting up the `cia402` comp module is kind of a pain, because of the number of pins that need to be wired up.  That's not the only source of pain, but it's probably the biggest for me so far.  So merging the logic from it into the CiA driver would be a usability win, *if* it makes sense from a code, etc perspective.
- The status and control words aren't *quite* just bitmapped, if I remember correctly.  Some of the bits change meaning depending on the mode, or interact in weird ways.  That's fine via an internal union like you've listed, as you still have access to the raw data.  It might even make sense to build per-mode structs into the union in some cases.  I'd have to re-read the spec and think about it a bit.
- I'm a bit hazy on how bitmapped structs and endianness interact.  We do endian conversion in when reading from PDOs, 
- We *probably* don't want to enforce having the driver handle the CiA state machine, as it reduces flexibility.  Maybe have an `enable` pin for that, defaulting to on?
- I haven't really thought about what adding a somewhat complex state machine to the code would end up doing to the structure of things.  We could *probably* just encapsulate it in a function that's only called if the state machine enable pin is set, but it'll need to clunk along for a while reading from various pins and writing state changes to the control word and maybe operation mode.
- Adding 3 pins for homing is no big deal.  The homing state machine is a bigger deal.  I don't really understand CiA 402 homing *or* LinuxCNC homing right now, and getting the two to interact correctly might end up being a pain.

Thinking about this, the first step is probably to get the basic mode-change state machine up and running, supporting CSP and CSV, implementing basically the same pins that the `cia402` component uses for talking to joints.  Then we can just swap the component out with a few minor renamings.

After that, we have a few options:

- Implement other "regular" modes.  I'd really like VL mode working soon, for spindles.  Getting CST (torque) might be interesting, but it probably won't get a lot of use.  I have hardware that supports it, at least.  And we *could* do PP, PV, and TQ, but I'm not sure how useful they'd be for EtherCAT devices that support CS* modes.  There are a handful of devices that don't support CS* modes, but they're probably CANopen devices with very thin EtherCAT shims around them.  I'm not really sure what would happen if you put a device into PP mode but fed it target updates every 1ms.  Would you get an approximation of CSP mode or a giant mess?  Easy enough to test, I guess.
- Implement homing.  Since homing is its own mode, I don't *think* it'd really interact deeply with the other modes in the state machine.  Assuming that we have to switch to HM mode, we'd need to save the old mode, transition to HM mode, deal with homing, finish homing, then switch back.  That's quite a few steps.

Also, how do users tell the state machine which mode they want to be in?  Writing into `opmode` isn't quite right, because there are states in that, too.  Using `cia402`-style binary pins isn't right either, when we have (potentially) 7 different motion modes.  If LinuxCNC has string pins, then I'd say to write "CSV" to one and be done with it.  Maybe an integer pin that matches opmode's values?  1 for PP, 3 for PV, 4 for TQ, 8 for CSP, etc?

Presumably we want to be able to support mode changes during operation?  For cases like threading with servo spindles?  I should review what HAL configs for that look like today.  Do people just leave servos in position mode when milling, or do they switch modes?  I'm trying to think of a CNC use case where spindle *speed* needs to be particularly accurate on a per-millisecond basis (vs spindle *position*, which you either don't care about at all or really really care a lot).

**Comment by rodw-au** [comment]
The bitmap is preserved but certain processes (eg switch on) has to manipulate some bits in a specific order. The rtelligent docs describe this pretty well.

endian is handled by the C compiler so an 8 bit bitmap follows the lo-byte/hi-byte order on x86 so nothing to do here as you have already done it.

Review my status union with Dominic's

```
  // read status
  stat_switchon_ready    = (statusword >> 0) & 1;
  stat_switched_on       = (statusword >> 1) & 1;
  stat_op_enabled        = (statusword >> 2) & 1;
  stat_fault             = (statusword >> 3) & 1;
  stat_voltage_enabled   = (statusword >> 4) & 1;
  stat_quick_stop        = (statusword >> 5) & 1;
  stat_switchon_disabled = (statusword >> 6) & 1;
  stat_warning           = (statusword >> 7) & 1;
  stat_remote            = (statusword >> 9) & 1;
```
ref: https://github.com/dbraun1981/hal-cia402/blob/main/cia402.comp#L216

cia402.comp only suports  CSV and CSP modes. Ref: https://github.com/dbraun1981/hal-cia402/blob/main/cia402.comp#L72
The enable pin sounds perfect. Maybe it fails if not in CSV and CSP modes initially (and expand modes later).
The additional logic from Dominic is really quite small. It switches the drive on, starts homing and reports errors. It probably also needs to "eat" following errors sent to linuxcnc during homing as linuxcnc knows nothing about the externally generated motion.

Homing is not a big deal. The driver just needs to start homing and report when its done. My homecomp will look after the rest (hopefully).

Due to the nature of current lcec/cia402 tools, I don't think modes will change after startup.

I hope I covered all your questions. 
This has some exciting potential.

**Comment by hmnijp** [comment]
I've been thinking about this, and I think that the lcec_cia402 driver should eventually absorb both cia402.comp and cia402_homecomp.

In setting up homemode for _abz encoder drives_, using [this component](https://github.com/marcoreps/linuxcnc_leadshine_EL8/blob/main/el8_homecomp.comp) as an example (Rodw used it as a basis for the cia402_homecomp he wants to do), I've found that I need to make changes both to the state machine programmed in cia402.comp (example changes are in[this thread](https://forum.linuxcnc.org/ethercat/44508-ethercat-installation-summary?start=30#282793) and [this one](https://forum.linuxcnc.org/ethercat/47358-kollmorgen-s300-danaher-s300-telemecanique-lexium-15-stuff-over-ethercat#279190)) and to the state machine in cia402_homecomp (similar to[these](https://forum.linuxcnc.org/ethercat/52968-help-needed-cracking-the-code-on-linuxcnc-servo-homing-setup?start=50)).  - Pay attention to these discussions. i can't give the final version as i haven't got it working properly yet...

The problem is that both components manipulate the homing procedure at the same time, but the resulting order is not quite right - this gives the wrong commands to the drive and causes jerks when switching opmode. 

This is complicated by the fact that I am using a CSV mod for this drive (closing the position loop inside linuxcnc using PID - in this mode the precision is much better than in csp mode and the position loop in the driver itself. maybe this is a feature of the Chinese driver - its position loop is poorly configured, and does not give a satisfactory result. maybe for other drives the csp mode will work acceptably, but the problem affects them too) - during opmode=6 linuxcnc thinks it is standing still but the drive is returning actual feedback during the homing operation, causing following error accumulation (linuxcnc estop status) and jerks of the motor during mode switching.
![2024-08-28 08-41-54](https://github.com/user-attachments/assets/89b891ad-acaf-4751-953a-6e2ccb621aef)
![2024-08-28 08-41-36](https://github.com/user-attachments/assets/5ead82c8-83e1-42fc-9073-210a5da60660)



Things are a bit easier with absolute encoders - it just uses a simplified state machine inside cia402.comp, in this case with HOME_ABSOLUTE_ENCODER = 2 it works fine - at startup linuxcnc just takes the position from the driver. 
But I haven't figured out how to simplify the process of setting the zero coordinates in the right place for this case (this needs to be done once between replacing the encoder battery) without doing it from the command line - it seems that this would require restarting linuxcnc by changing the HOME_ABSOLUTE_ENCODER = 0 settings, running opmode=6, then returning cnc.ini to the previous state - probably this process should be coded into cia402 as well. 


This would all be easier to tune if the whole algorithm was in one place, instead of fixing basic_cia402, cia402, and cia402homemode

**Comment by rodw-au** [comment]
I agree with you here. cia402.comp was written by dominic and tested with  absolute encoders hence other things are not working well.  I once got homing and cia402 nearly working on the one homecomp. https://github.com/rodw-au/ciahome Then the homecomp template changed and obsoleted my work so I lost interest. (its a lot easier today with the changes). My code does include a nice clean interface to manage the ethercat statusword and controlword using a union structure to slice out the bits.

I began to think that the homecomp is not the right place for the cia402 state machine because being an internal Linuxcnc module, homecomp sits outside of the servo-thread's lcec loop. I though this might mean that the cia402.comp was running one servo thread cycle behind if it is in the homecomp.

I have finally got  a bit of time to look at this and have my ethercat homing lab set up
![ethercat-lab](https://github.com/user-attachments/assets/394ac322-65f4-43fb-9734-b6c10419bb75)

I have not been able to solve an internal drive error when motion starts on the closed loop motor but it works in open loop. This is frustrating becasue I did have it running in the past.

My plan is to run in open loop which should still allow me to home to a sensor but probably not allow me to home via torque sensing.

After that, I was going to look at how to incorporate the cia402 states into lcec_cia402.
I suspect this should be easier becasue there are seperate instances of the driver for each slave. In the homecomp, you have to manage all instances in the one procedure which makes the code much more complicated.

So I am all in with this!

**Reply by hmnijp** [reply]
> I began to think that the homecomp is not the right place for the cia402 state machine because being an internal Linuxcnc module, homecomp sits outside of the servo-thread's lcec loop. I though this might mean that the cia402.comp was running one servo thread cycle behind if it is in the homecomp.

yes, due to the many connections between components, from my graphs I see that the entire control return cycle takes more than 6 servo cycles...


> I have not been able to solve an internal drive error when motion starts on the closed loop motor but it works in open loop. This is frustrating becasue I did have it running in the past.

yes, after switching there is an error related to the incorrectness of the Motor-pos-cmd command, even without a closed loop. So far I have had little time to track the entire sequence of actions, and where this offset comes from.

a closed loop adds to this a slow reaction of the pid which calculates the error from the previous cycle and gives an additional incorrect signal for regulation.
I think this should be solved by connecting the connection
=> pid.index-enable

maybe soon i will take servos for homework so i will have more time to test it to find the error...
I have the opportunity to conduct tests on Lichuan lc10e servos, lichuan stepper cl3, leadshine el6 servos with absolute and increment encoders

**Comment by hmnijp** [comment]
Also, I have not seen any ethercat config examples of using PID in Hal - which is a prerequisite for control in CSV mode, without it the motion will be completely uncontrollable - following error will accumulate every second, causing significant deviations. This must be corrected using PID and position feedback. This configuration is quite simple, and it is necessary to supplement this configuration examples for `basic_cia402` so that beginners do not have misunderstandings.


> Presumably we want to be able to support mode changes during operation? For cases like threading with servo spindles? I should review what HAL configs for that look like today. Do people just leave servos in position mode when milling, or do they switch modes? I'm trying to think of a CNC use case where spindle speed needs to be particularly accurate on a per-millisecond basis (vs spindle position, which you either don't care about at all or really really care a lot).

As for switching from `csv` to `csp` mode, this also works and can be used. Just like a default configuration LinuxCNC with MESA board, you can switch between velocity and position control while running by changing the state:
`setp hm2.stepgen.00.control-type = 0/1` (velocity or position)
control is switched from:
`joint.0.motor-pos-cmd => hm2.stepgen.00.position-cmd`
to:
`joint.0.vel-cmd => {pid.command => pid.output} => hm2.stepgen.00.velocity-cmd`
`hm2.stepgen.00.position-fb => joint.0.motor-pos-fb & pid.x.feedback`

the same way cia402 performs state switching:
`setp cia402.0.csp-mode = 0/1`  (velocity or position, switch opmode 8 to 9)
control is switched from:
`joint.0.motor-pos-cmd => cia402.0.pos-cmd`
to:
`joint.0.vel-cmd => { pid.command => pid.output} => cia402.0.velocity-cmd`
`cia402.0.pos-fb => joint.0.motor-pos-fb & pid.x.feedback`

![2024-08-28 09-47-26](https://github.com/user-attachments/assets/2dd7472e-b34f-4901-b61a-04620ca3efee)



As you can see, even with a simple Mesa card and stepdir control, the following error increases significantly when switching from CV to CP. And with the popularity of many low-cost Chinese servos that do not have a good auto-tuning algorithm, many people will have to experiment and perform the mode change. Doing this change opmode at runtime will help you to see the difference in the amount of following error (using the halscope graph), and to properly set the pid loop in the driver and linuxcnc.
As I mentioned before, the china servos I tested did not satisfy me in CSP mode.

There are also more complex examples - for example millturn ( [like this ](https://www.youtube.com/watch?v=kWfjtIbxvdM) or [site](http://www.talla83.de/linuxcnc/config.htm)), where in hal the servo control is switched between the 
`spindle.0.speed-out`
to
`joint.C-axis.motor-pos-cmd`
...and many Nets to provide home reset between switching - the joint thinks it is not moving and forgets its position while the drive is controlled by velocity from spindle.comp

> For cases like threading with servo spindles? 

Threading configuration with a spindle is complicated by the fact that linuxcnc does not use the spindle position. It requires an `index-enable` signal and `encoder reset` before synchronous axis motion can be started move, and then the linear axis will move synchronously maintaining equal velocity with the spindle. 
https://linuxcnc.org/docs/html/gcode/g-code.html#gcode:g33
https://linuxcnc.org/docs/html/examples/spindle.html#_spindle_synchronized_motion

(I think this obsolete code is long overdue for revision - it started with simple lathes using only 1 or a few index on the spindle and a spindle speed command. But precision encoders allow position feedback to be used for synchronization instead of the encoder index)

> I should review what HAL configs for that look like today. Do people just leave servos in position mode when milling, or do they switch modes?

If you don't have much hands-on experience with the nuances of using linuxcnc, I can help and provide a complete example config including all ini parameters and hal pins for a basic example. (similar to the config from the pncconf wizard). This will make it much easier for newbies who don't know much about Hal and Ethercat to get started.
